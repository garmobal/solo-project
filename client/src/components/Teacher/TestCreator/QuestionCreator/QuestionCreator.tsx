import React, { useState } from 'react';
import CreateButton from '../../../UI/CreateButton/CreateButton';

import styles from './QuestionCreator.module.scss';

interface IProps {
  saveQuestion: Function;
}

const QuestionCreator = (props: IProps) => {
  const [titleValue, setTitleValue] = useState('');

  const inputTitleHandler = (e: React.ChangeEvent<HTMLInputElement>) =>
    setTitleValue(e.target.value);

  const onSubmitHandler = (e: any) => {
    e.preventDefault();
    props.saveQuestion(e.target);
    setTitleValue('');
  };

  return (
    <div className={styles.QuestionCreator}>
      <form onSubmit={onSubmitHandler}>
        <input
          type="text"
          name="question"
          aria-label="question"
          placeholder="Question"
          autoComplete="off"
          required
          value={titleValue}
          onChange={inputTitleHandler}
        />
        <input
          type="text"
          name="a"
          aria-label="option-a"
          placeholder="Option a"
          autoComplete="off"
          required
        />
        <input
          type="text"
          name="b"
          aria-label="option-b"
          placeholder="Option b"
          autoComplete="off"
          required
        />
        <input
          type="text"
          name="c"
          aria-label="option-c"
          placeholder="Option c"
          autoComplete="off"
          required
        />
        <input
          type="text"
          name="d"
          aria-label="option-d"
          placeholder="Option d"
          autoComplete="off"
          required
        />
        <select aria-label="correctanswer" name="answer">
          <option value="a">a</option>
          <option value="b">b</option>
          <option value="c">c</option>
          <option value="d">d</option>
        </select>
        <input
          type="file"
          placeholder="Choose image"
          name="selectedImage"
          className={styles.ImageInput}
        />
        <div className={styles.Button}>
          <CreateButton type={'submit'}>
            <div>Add Question</div>
          </CreateButton>
        </div>
      </form>
    </div>
  );
};

export default QuestionCreator;
