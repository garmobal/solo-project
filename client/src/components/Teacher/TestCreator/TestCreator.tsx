import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { History } from 'history';
import CreateButton from '../../UI/CreateButton/CreateButton';
import QuestionCard from './QuestionCard/QuestionCard';
import QuestionCreator from './QuestionCreator/QuestionCreator';
import { postTest } from '../../../store/actions/testActions';

import styles from './TestCreator.module.scss';
import { TestQuestion } from '../../../types';

interface IProps {
  history: History;
  
}

const TestCreator = (props: IProps) => {
  const [questions, setQuestions] = useState<TestQuestion[]>([]);
  const [title, setTitle] = useState('Test 1');
  const [isValid, setIsValid] = useState(true);

  useEffect(() => {
    // Retrieve questions from storage
    const item = JSON.parse(localStorage.getItem('currenttest') || '{}');
    if (!item) localStorage.setItem('currenttest', JSON.stringify({}));
    else {
      const storedQuestions = [];
      for (let key of Object.keys(item)) {
        storedQuestions.push(item[key]);
      }
      if (storedQuestions.length) setQuestions(storedQuestions);
    }
  }, []);

  const saveQuestionHandler = (q: React.ChangeEvent<HTMLInputElement>) => {
    const question = q as typeof q & {
      [key: string]: { value: string };
      question: { value: string };
      answer: { value: string };
    };
    const questionFile = q as typeof q & {
      selectedImage: { files: File[] };
    };
    // Create question object
    const letters = ['a', 'b', 'c', 'd'];

    const options = letters.map((ch) => ({
      op: question[ch].value,
      correct: false,
    }));
    const newQuestion: TestQuestion = {
      question: question.question.value,
      options,
      answer: question[question.answer.value].value,
    };
    newQuestion.options.map(
      (opt) => (opt.correct = opt.op === newQuestion.answer)
    );

    if (questionFile.selectedImage.files[0]) {
      // SHOW SPINNER
      const data = new FormData();
      data.append('file', questionFile.selectedImage.files[0]);
      data.append('upload_preset', 'learntoday');

      fetch('https://api.cloudinary.com/v1_1/garmobal/image/upload', {
        method: 'POST',
        body: data,
      })
        .then((res) => res.json())
        .then((res) => {
          newQuestion.image = res.secure_url;
          setQuestions((currentQ) => [...currentQ, newQuestion]);
          const currenttest = JSON.parse(
            localStorage.getItem('currenttest') || '{}'
          );
          currenttest[newQuestion.question] = newQuestion;
          localStorage.setItem('currenttest', JSON.stringify(currenttest));
          //HIDE SPINNER
        });
    } else {
      setQuestions((currentQ) => [...currentQ, newQuestion]);
      const currenttest = JSON.parse(
        localStorage.getItem('currenttest') || '{}'
      );
      currenttest[newQuestion.question] = newQuestion;
      localStorage.setItem('currenttest', JSON.stringify(currenttest));
    }
  };

  const trashHandler = (question: string) => {
    const currenttest = JSON.parse(localStorage.getItem('currenttest') || '{}');
    delete currenttest[question];
    localStorage.setItem('currenttest', JSON.stringify(currenttest));
    setQuestions(questions.filter((q) => q.question !== question));
  };

  const titleHandler = (e: React.ChangeEvent<HTMLInputElement>) =>
    setTitle(e.target.value);

  const dispatch = useDispatch();

  const submitTestHandler = () => {
    if (questions.length < 4) setIsValid(false);
    else {
      dispatch(postTest(questions, title));
      localStorage.clear();
      props.history.replace('/tests');
    }
  };

  const addMoreQuestionsMessage = (
    <p className={isValid ? styles.Hidden : styles.Warning}>
      Tests must have at least 4 questions and a title
    </p>
  );

  return (
    <div className={styles.TestCreator}>
      <div className={styles.QuestionsList}>
        {questions.length ? (
          questions.map((q, i) => (
            <QuestionCard key={i} quest={q} trashHandler={trashHandler} />
          ))
        ) : (
          <p className={styles.Message}>Start by adding some questions!</p>
        )}
      </div>
      <QuestionCreator saveQuestion={saveQuestionHandler} />
      <hr className={styles.LineBreak} />
      <div className={styles.SaveTest}>
        <input
          type="text"
          aria-label="test-title"
          placeholder="Add a test title"
          onChange={titleHandler}
          className={styles.Title}
        />
        <div className={styles.TestButtonDiv}>
          <CreateButton clicked={submitTestHandler}>
            <div>Save Test</div>
          </CreateButton>
        </div>
        {questions.length < 4 ? addMoreQuestionsMessage : null}
      </div>
    </div>
  );
};

export default TestCreator;
