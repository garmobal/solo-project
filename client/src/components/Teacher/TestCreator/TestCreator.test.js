import React from 'react';

import { fireEvent, render, screen } from '../../../__tests__/test-utils';
import QuestionCreator from './QuestionCreator/QuestionCreator';
import TestCreator from './TestCreator.tsx';

const setup = () => {
  const utilsquestion = render(<TestCreator />);
  const question = utilsquestion.getByLabelText('question');
  const optiona = utilsquestion.getByLabelText('option-a');
  const optionb = utilsquestion.getByLabelText('option-b');
  const optionc = utilsquestion.getByLabelText('option-c');
  const optiond = utilsquestion.getByLabelText('option-d');
  const correctanswer = utilsquestion.getByLabelText('correctanswer');
  return {
    question,
    optiona,
    optionb,
    optionc,
    optiond,
    correctanswer,
    ...utilsquestion,
  };
};

describe('TestCreator', () => {
  it('should render the page', () => {
    render(<TestCreator />);
    expect(
      screen.getByText(/Start by adding some questions!/i)
    ).toBeInTheDocument();
  });
  it('should be able to create a question', () => {
    const onSubmit = jest.fn();
    const {
      question,
      optiona,
      optionb,
      optionc,
      optiond,
      correctanswer,
    } = setup();
    fireEvent.change(question, {
      target: { question: { value: 'RandomQuestion' } },
    });
    fireEvent.change(optiona, { target: { value: 'RandomAnswer' } });
    fireEvent.change(optionb, { target: { value: 'RandomAnswer' } });
    fireEvent.change(optionc, { target: { value: 'RandomAnswer' } });
    fireEvent.change(optiond, { target: { value: 'RandomAnswer' } });
    fireEvent.change(correctanswer, { target: { value: 'a' } });
    fireEvent.click(screen.getByText(/Add Question/i));
    expect(onSubmit).toBeCalled();
  });
});
