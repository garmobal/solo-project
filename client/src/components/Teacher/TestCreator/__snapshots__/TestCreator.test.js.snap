// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders successfully 1`] = `ShallowWrapper {}`;
